src/plugins/Miner.cpp

```diff
@@ -44,7 +44,12 @@ void SecureVespeneIncome() {
     auto refineries = gAPI->observer().GetUnits(IsRefinery());
 
     for (const auto& i : refineries()) {
-        if (i->assigned_harvesters >= i->ideal_harvesters)
+        if (i->assigned_harvesters > i->ideal_harvesters) {
+            gHub->RemoveVespeneHarvester(*i);
```

Comment:
Actually, I think the problem is in the assignment logic. How it happens that we have extra harvesters?

Reply:
The SCV that builds the Refinery takes 5+ frames before it exits the 'geyser' construction area and turns around and joins the refinery. So for Terran even with perfect-working assignment, there will be 4/3. Unless we keep a separate list of vespene-workers-per-refinery.

Other than that, I think it takes ~5 frames from when the building is finished to when the "assigned_harvesters" shows a valid value. It's possible it doesn't check every frame to update "assigned_harvesters" as I have seen up to 8/3 for a geyser, which makes me think it checks every 5 frames, so if assigning the 3rd vespener on frame 1000, it will still show 2/3 (1 frame action-lag) until frame 1005, when it updates to 8/3 (including the builder).

Reply:
I think, the first mitigation: we can manually assign that building scv as harvester. 

The second could be making miner trigger once in 6 or more frames. So this lag could be skipped.

The third, I think is to ignore the data, provided by the game and get list of the assigned harvesters from the cache somehow.

Reply:
Just running SecureVespene only every 10 frames does fix this problem completely. Fix is very simple now. :)